src/Courses/TopicGrades/GradeInfoHeader.tsx

```diff
@@ -42,154 +41,260 @@ const useStyles = makeStyles((theme) => ({
+        } else if (_.isNil(grade.workbooks) || _.isNil(info.workbookId)) {
+            // no workbooks, or none selected -> use currentProblemState (need version info if an exam... grade instance ID will not suffice)
+            if (!_.isNil(info.studentGradeInstanceId) && !_.isNil(grade.gradeInstances) && !_.isEmpty(grade.gradeInstances)) {
+                newProblemState.studentTopicAssessmentInfoId = _.find(grade.gradeInstances, ['id', info.studentGradeInstanceId])?.studentTopicAssessmentInfoId;
+            }
+        } else if (_.isNil(grade.workbooks[info.workbookId])) {
+            // we have workbooks, and one is selected, but entry doesn't exist -> error
+            logger.error(`GradeInfoHeader: User #${selected.user?.id} tried to set workbook #${info.workbookId} for problem #${selected.problem?.id} but I cannot find that record.`);
+        } else { 
+            newProblemState.workbookId = info.workbookId;
+        }
+
+        const newGradeInstance = (info.studentGradeInstanceId) ? _.find(grade?.gradeInstances, ['id', info.studentGradeInstanceId]) : undefined;
+        logger.debug('GradeInfoHeader: setting new selected grades', newProblemState, newGradeInstance);
+        setSelected(selected => ({ ...selected, problemState: newProblemState, grade: currentGrade, gradeInstance: newGradeInstance }));
+    }, [info.workbookId, info.studentGradeId, info.studentGradeInstanceId]);
 
     const fetchTopicGrade = async () => {
         setTopicGrade(null);
-        if (_.isNil(topicId)) {
-            logger.debug('GradeInfoHeader topicId id is nil, skipping grades call');
+        if (_.isNil(topic)) {
+            logger.debug('GradeInfoHeader: topic id is nil, skipping grades call');
         }
 
         const params = {
-            topicId: topicId,
-            userId: grade.userId
+            topicId: topic.id,
+            userId: selected.user?.id // check selected.user not nil?
         };
         const result = await getGrades(params);
         setTopicGrade(result.data.data.first?.average ?? null);
     };
     useEffect(() => {
+        logger.debug('GradeInfoHeader: there has been a change in user or topic, fetching new topic grade.');
         fetchTopicGrade();
-    }, [grade.userId, topicId]);
+    }, [selected.user, topic]);
 
-    const workbookList = (vMap: Record<number, Array<number>>, versioned: boolean) => {
-        const options: WorkbookOption[] = [];
-        const versionKeys = _.keys(vMap).map((v) => { return parseInt(v, 10); }).sort();
+    const onSuccess = (gradeOverride: Partial<StudentGrade>) => {
+        logger.debug('GradeInfoHeader: overriding grade', gradeOverride.effectiveScore);
+        const currentGrade = grade; 
+        if (!_.isNil(gradeOverride) &&
+            !_.isNil(gradeOverride.effectiveScore) &&
+            !_.isNil(currentGrade)
+        ) {
+            currentGrade.effectiveScore = gradeOverride.effectiveScore;
+            setGrade(currentGrade);
```

Comment:
This shouldn't trigger a rerender since it is the same object

Reply:
not sure what you're saying - but this does update the displayed effective score when an override is made...

Reply:
Gotcha, there is probably something else triggering the rerender

State only calls render when the reference changes, so for primitives like numbers it's whenever the value changes

However when you setState with an object and you set it to itself it does not trigger a rerender

That is why in a lot of places you'll see stuff like:
```
setTopic(new Topic(currentTopic))
```
or
```
setTopic({...topic})
```

Those are different ways of updating the reference

Reply:
Maybe add a TODO to investigate in the future